builder/packages_image.go

```diff
@@ -39,86 +40,135 @@ func NewPackagesImageBuilder(repository, compiledPackagesPath, targetPath, fissi
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+		if _, err = io.CopyN(w.stream, file, info.Size()); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	// Generate configuration
-	p.ui.Println("Generating configuration JSON specs ...")
-	configStore := configstore.NewConfigStoreBuilder(
-		configstore.JSONProvider,
-		lightManifestPath,
-		darkManifestPath,
-		filepath.Join(rootDir, "opt/hcf/specs"),
-	)
-
-	if err := configStore.WriteBaseConfig(roleManifest); err != nil {
-		return "", fmt.Errorf("Error writing base config: %s", err.Error())
+// CreatePackagesDockerStream generates a tar stream containing the docker context to build the packages layer image with
+func (p *PackagesImageBuilder) CreatePackagesDockerStream(roleManifest *model.RoleManifest, lightManifestPath, darkManifestPath string) (io.ReadCloser, <-chan error, error) {
+	if len(roleManifest.Roles) == 0 {
+		return nil, nil, fmt.Errorf("No roles to build")
 	}
 
-	// Copy packages
-	p.ui.Println("Copying compiled packages...")
-	copiedFingerprints := make(map[string]bool)
-	for _, role := range roleManifest.Roles {
-		for _, job := range role.Jobs {
-			for _, pkg := range job.Packages {
-				if _, ok := copiedFingerprints[pkg.Fingerprint]; ok {
-					// Package has already been copied (possibly due to a different role)
-					continue
-				}
+	pipeReader, pipeWriter := io.Pipe()
+	errors := make(chan error)
 
-				compiledDir := pkg.GetPackageCompiledDir(p.compiledPackagesPath)
-				if err := util.ValidatePath(compiledDir, true, fmt.Sprintf("compiled dir for package %s", pkg.Name)); err != nil {
-					return "", err
-				}
+	go func() {
+		defer close(errors)
+		defer pipeWriter.Close()
+		tarStream := tar.NewWriter(pipeWriter)
+		defer tarStream.Close()
 
-				packageDir := filepath.Join(rootDir, "var/vcap/packages-src", pkg.Fingerprint)
-				err = shutil.CopyTree(
-					compiledDir,
-					packageDir,
-					&shutil.CopyTreeOptions{
-						Symlinks:               true,
-						Ignore:                 nil,
-						CopyFunction:           shutil.Copy,
-						IgnoreDanglingSymlinks: false,
-					},
-				)
-				if err != nil {
-					return "", err
-				}
+		// Generate configuration
+		specDir, err := ioutil.TempDir("", "fissile-spec-dir")
+		if err != nil {
+			errors <- err
+			return
+		}
+		defer os.RemoveAll(specDir)
+		configStore := configstore.NewConfigStoreBuilder(
+			configstore.JSONProvider,
+			lightManifestPath,
+			darkManifestPath,
+			specDir,
+		)
+
+		if err := configStore.WriteBaseConfig(roleManifest); err != nil {
+			errors <- fmt.Errorf("Error writing base config: %s", err.Error())
+			return
+		}
+		walker := &tarWalker{stream: tarStream, root: specDir, prefix: "specs"}
+		if err := filepath.Walk(specDir, walker.walk); err != nil {
+			errors <- err
+			return
+		}
 
-				copiedFingerprints[pkg.Fingerprint] = true
+		// Collect compiled packages
+		foundFingerprints := make(map[string]bool)
```

Comment:
Typically when using a map as a set people will use an empty struct for a zero allocation type of thing: `map[string]struct{}`. This can generally be used to replace most uses of `map[string]bool`. It's not strictly necessary but just letting you know.